maths/modular_inverse.nim

```diff
@@ -0,0 +1,65 @@
+## Modular inverse
+
+import std/[options, math]
+
+
+func euclidHalfIteration(inA, inB: Positive): (Natural, int) =
+  var (a, b) = (inA.Natural, inB.Natural)
+  var (x0, x1) = (1, 0)
+  while b != 0:
+    (x0, x1) = (x1, x0 - (a div b) * x1)
+    (a, b) = (b, a mod b)
```

Comment:
```suggestion
    (a, b) = (b, math.floorMod(a, b))
```
I think that math.floorMod might be more suited here than Nim's mod operator which may return a negative integer.
You can then return two Naturals.
Another maybe less expensive trick to ensure that `x0` is positive, is to perform an addition in the end:
(x0 + a mod a)

Reply:
I am not sure what do you mean by `(x0 + a mod a)`. In this situation, `a mod a` is always `0`.

Reply:
I meant $(x_0 + a) \bmod{a}$.

Ensuring that `x0` is positive enables using unsigned arithmetic for large modulus.
The `mod` operator's behaviour is different from the `%` C operator. It returns a result with the sign of its operands which are non-negative in this case.
I would rather regroup everything under the same function `modInvByEuclidsXGCD` and put a comment to distinguish the part that is similar to Euclid's algorithm.
Half-iteration is not a commonly accepted denomination, nor a different algorithm.

Reply:
> Ensuring that `x0` is positive enables using unsigned arithmetic for large modulus.

During the computation /_iteration_ `x0` still might become negative. Or am I missing something?

> The `mod` operator's behavior is different from the `%` C operator. It returns a result with the sign of its operands which are non-negative in this case.

There is a [good wiki page](https://en.wikipedia.org/wiki/Modulo#In_programming_languages) about that. Basically I _do not trust_ `mod` for negative inputs.

> I would rather regroup everything under the same function `modInvByEuclidsXGCD` and put a comment to distinguish the part that is similar to Euclid's algorithm. Half-iteration is not a commonly accepted denomination, nor a different algorithm.

I prefer to have a separate functions: `euclidHalfIteration` does the computation, `modularInverse*` handles all of the sign madness. This provides more structure than the comments. Similarly as discussed in #62, `euclidHalfIteration` is not a perfect name.

Reply:
I changed my mind. Keep it in two separate functions, `euclidHalfIteration` should be fine.

> During the computation /iteration x0 still might become negative
Could you let me know if you are sure about this? Can you give me two Positive inputs, for which the result of mod is negative?
Rather than the wiki, I checked Nim's official mod documentation in the system module. It says that the result of two positive inputs (quotient and dividend are positive) should return a positive remainder.

Reply:
If you compute `modularInverse(3, 97)` at some point you will reach the state with `x0 = -32`.